File: cpp/src/hypo.cpp
#include "hypo.h"

// JSON Keys
#define LATITUDE_KEY "Latitude"
#define LONGITUDE_KEY "Longitude"
#define DEPTH_KEY "Depth"
#define TIME_KEY "Time"
#define LATITUDE_ERROR_KEY "LatitudeError"
#define LONGITUDE_ERROR_KEY "LongitudeError"
#define DEPTH_ERROR_KEY "DepthError"
#define TIME_ERROR_KEY "TimeError"

namespace detectionformats {

hypo::hypo() {
	latitude = std::numeric_limits<double>::quiet_NaN();
	longitude = std::numeric_limits<double>::quiet_NaN();
	depth = std::numeric_limits<double>::quiet_NaN();
	time = std::numeric_limits<double>::quiet_NaN();
	latitudeerror = std::numeric_limits<double>::quiet_NaN();
	longitudeerror = std::numeric_limits<double>::quiet_NaN();
	deptherror = std::numeric_limits<double>::quiet_NaN();
	timeerror = std::numeric_limits<double>::quiet_NaN();

}

hypo::hypo(double newlatitude, double newlongitude, double newtime,
		double newdepth, double newlatitudeerror, double newlongitudeerror,
		double newtimeerror, double newdeptherror) {
	latitude = newlatitude;
	longitude = newlongitude;
	depth = newdepth;
	time = newtime;
	latitudeerror = newlatitudeerror;
	longitudeerror = newlongitudeerror;
	deptherror = newdeptherror;
	timeerror = newtimeerror;
}

hypo::hypo(rapidjson::Value &json) {
	// required values
	// latitude
	if ((json.HasMember(LATITUDE_KEY) == true)
			&& (json[LATITUDE_KEY].IsNumber() == true)
			&& (json[LATITUDE_KEY].IsDouble() == true))
		latitude = json[LATITUDE_KEY].GetDouble();
	else
		latitude = std::numeric_limits<double>::quiet_NaN();

	// longitude
	if ((json.HasMember(LONGITUDE_KEY) == true)
			&& (json[LONGITUDE_KEY].IsNumber() == true)
			&& (json[LONGITUDE_KEY].IsDouble() == true))
		longitude = json[LONGITUDE_KEY].GetDouble();
	else
		longitude = std::numeric_limits<double>::quiet_NaN();

	// time
	if ((json.HasMember(TIME_KEY) == true)
			&& (json[TIME_KEY].IsString() == true))
		time = detectionformats::ConvertISO8601ToEpochTime(
				std::string(json[TIME_KEY].GetString(),
						json[TIME_KEY].GetStringLength()));
	else
		time = std::numeric_limits<double>::quiet_NaN();

	// depth
	if ((json.HasMember(DEPTH_KEY) == true)
			&& (json[DEPTH_KEY].IsNumber() == true)
			&& (json[DEPTH_KEY].IsDouble() == true))
		depth = json[DEPTH_KEY].GetDouble();
	else
		depth = std::numeric_limits<double>::quiet_NaN();

	// optional values
	// latitude error
	if ((json.HasMember(LATITUDE_ERROR_KEY) == true)
			&& (json[LATITUDE_ERROR_KEY].IsNumber() == true)
			&& (json[LATITUDE_ERROR_KEY].IsDouble() == true))
		latitudeerror = json[LATITUDE_ERROR_KEY].GetDouble();
	else
		latitudeerror = std::numeric_limits<double>::quiet_NaN();

	// longitude error
	if ((json.HasMember(LONGITUDE_ERROR_KEY) == true)
			&& (json[LONGITUDE_ERROR_KEY].IsNumber() == true)
			&& (json[LONGITUDE_ERROR_KEY].IsDouble() == true))
		longitudeerror = json[LONGITUDE_ERROR_KEY].GetDouble();
	else
		longitudeerror = std::numeric_limits<double>::quiet_NaN();

	// time error
	if ((json.HasMember(TIME_ERROR_KEY) == true)
			&& (json[TIME_ERROR_KEY].IsNumber() == true)
			&& (json[TIME_ERROR_KEY].IsDouble() == true))
		timeerror = json[TIME_ERROR_KEY].GetDouble();
	else
		timeerror = std::numeric_limits<double>::quiet_NaN();

	// depth error
	if ((json.HasMember(DEPTH_ERROR_KEY) == true)
			&& (json[DEPTH_ERROR_KEY].IsNumber() == true)
			&& (json[DEPTH_ERROR_KEY].IsDouble() == true))
		deptherror = json[DEPTH_ERROR_KEY].GetDouble();
	else
		deptherror = std::numeric_limits<double>::quiet_NaN();
}

hypo::hypo(const hypo & newhypo) {
	latitude = newhypo.latitude;
	longitude = newhypo.longitude;
	depth = newhypo.depth;
	time = newhypo.time;
	latitudeerror = newhypo.latitudeerror;
	longitudeerror = newhypo.longitudeerror;
	deptherror = newhypo.deptherror;
	timeerror = newhypo.timeerror;
}

hypo::~hypo() {
}

rapidjson::Value & hypo::tojson(rapidjson::Value &json,
		rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> &allocator) {
	json.SetObject();

	// required values
	// latitude
	if (std::isnan(latitude) != true)
		json.AddMember(LATITUDE_KEY, latitude, allocator);

	// longitude
	if (std::isnan(longitude) != true)
		json.AddMember(LONGITUDE_KEY, longitude, allocator);

	// time
	if (std::isnan(time) != true) {
		std::string timestring = detectionformats::ConvertEpochTimeToISO8601(
				time);
		rapidjson::Value timevalue;
		timevalue.SetString(rapidjson::StringRef(timestring.c_str()),
				allocator);
		json.AddMember(TIME_KEY, timevalue, allocator);
	}

	// depth
	if (std::isnan(depth) != true)
		json.AddMember(DEPTH_KEY, depth, allocator);

	// optional values
	// latitude error
	if (std::isnan(latitudeerror) != true)
		json.AddMember(LATITUDE_ERROR_KEY, latitudeerror, allocator);

	// longitude error
	if (std::isnan(longitudeerror) != true)
		json.AddMember(LONGITUDE_ERROR_KEY, longitudeerror, allocator);

	// time error
	if (std::isnan(timeerror) != true)
		json.AddMember(TIME_ERROR_KEY, timeerror, allocator);

	// depth error
	if (std::isnan(deptherror) != true)
		json.AddMember(DEPTH_ERROR_KEY, deptherror, allocator);

	return (json);
}

std::vector<std::string> hypo::geterrors() {
	std::vector<std::string> errorlist;

	// check required data
	// latitude
	if (std::isnan(latitude) == true) {
		// latitude not found
		errorlist.push_back("No Latitude in hypo class.");
	}
	if ((latitude < -90) || (latitude > 90)) {
		errorlist.push_back("Invalid Latitude in hypo class.");
	}

	// longitude
	if (std::isnan(longitude) == true) {
		// longitude not found
		errorlist.push_back("No Longitude in hypo class.");
	}
	if ((longitude < -180) || (longitude > 180)) {
		errorlist.push_back("Invalid Longitude in hypo class.");
	}

	// time
	if (std::isnan(time) == true) {
		errorlist.push_back("Time is missing in hypo class.");
	} else {
		try {
			if (detectionformats::IsStringISO8601(
					detectionformats::ConvertEpochTimeToISO8601(time))
					== false) {
				errorlist.push_back("Time did not validate in hypo class.");
			}
		} catch (const std::exception & e) {
			errorlist.push_back(std::string(e.what()));
		}
	}

	// depth
	if (std::isnan(depth) == true) {
		// depth not found
		errorlist.push_back("No Depth in hypo class.");
	}
	if ((depth < -100) || (depth > 1500)) {
		errorlist.push_back("Invalid Depth in hypo class.");
	}

	// optional keys
	// Currently no validation criteria for optional values LatitudeError,
	// LongitudeError, TimeError, and DepthError.

	// return the list of errors
	return (errorlist);
}

}

File: java/src/gov/usgs/detectionformats/Hypo.java
package gov.usgs.detectionformats;

import java.util.ArrayList;
import java.util.Date;

import org.json.simple.JSONObject;

/**
 * a conversion class used to create, parse, and validate hypocenter data as
 * part of detection data.
 * 
 * @author U.S. Geological Survey &lt;jpatton at usgs.gov&gt;
 */
public class Hypo implements DetectionInt {

	/**
	 * JSON Keys
	 */
	public static final String LATITUDE_KEY = "Latitude";
	public static final String LONGITUDE_KEY = "Longitude";
	public static final String DEPTH_KEY = "Depth";
	public static final String TIME_KEY = "Time";
	public static final String LATITUDE_ERROR_KEY = "LatitudeError";
	public static final String LONGITUDE_ERROR_KEY = "LongitudeError";
	public static final String DEPTH_ERROR_KEY = "DepthError";
	public static final String TIME_ERROR_KEY = "TimeError";

	/**
	 * Required Double containing the latitude
	 */
	private final Double latitude;

	/**
	 * Required Double containing the longitude
	 */
	private final Double longitude;

	/**
	 * Required time for this hypo
	 */
	private final Date time;

	/**
	 * Required Double containing the depth
	 */
	private final Double depth;

	/**
	 * Optional Double containing the latitude error
	 */
	private final Double latitudeError;

	/**
	 * Optional Double containing the longitude error
	 */
	private final Double longitudeError;

	/**
	 * Optional Double containing the time error
	 */
	private final Double timeError;

	/**
	 * Optional Double containing the depth error
	 */
	private final Double depthError;

	/**
	 * The constructor for the Hypo class. Initializes members to null values.
	 */
	public Hypo() {

		latitude = null;
		longitude = null;
		time = null;
		depth = null;
		latitudeError = null;
		longitudeError = null;
		timeError = null;
		depthError = null;
	}

	/**
	 * Advanced constructor
	 * 
	 * The alternate advanced constructor for the Origin class. Initializes
	 * members to provided values.
	 * 
	 * @param newLatitude
	 *            - A Double containing the latitude to use
	 * @param newLongitude
	 *            - A Double containing the longitude to use
	 * @param newTime
	 *            - A Date containing the new origin time to use
	 * @param newDepth
	 *            - A Double containing the depth to use
	 * @param newLatitudeError
	 *            - A Double containing the latitude error to use, null to omit
	 * @param newLongitudeError
	 *            - A Double containing the longitude error to use, null to omit
	 * @param newTimeError
	 *            - A Double containing the new time error to use, null to omit
	 * @param newDepthError
	 *            - A Double containing the depth error to use, null to omit
	 */
	public Hypo(Double newLatitude, Double newLongitude, Date newTime,
			Double newDepth, Double newLatitudeError, Double newLongitudeError,
			double newTimeError, Double newDepthError) {

		latitude = newLatitude;
		longitude = newLongitude;
		time = newTime;
		depth = newDepth;
		latitudeError = newLatitudeError;
		longitudeError = newLongitudeError;
		timeError = newTimeError;
		depthError = newDepthError;
	}

	/**
	 * Constructs the class from a JSONObject, populating members
	 * 
	 * @param newJSONObject
	 *            - A JSONObject.
	 */
	public Hypo(JSONObject newJSONObject) {

		// Required values
		// latitude
		if (newJSONObject.containsKey(LATITUDE_KEY)) {
			latitude = (double) newJSONObject.get(LATITUDE_KEY);
		} else {
			latitude = null;
		}

		// longitude
		if (newJSONObject.containsKey(LONGITUDE_KEY)) {
			longitude = (double) newJSONObject.get(LONGITUDE_KEY);
		} else {
			longitude = null;
		}

		// time
		if (newJSONObject.containsKey(TIME_KEY)) {
			time = Utility.getDate(newJSONObject.get(TIME_KEY).toString());
		} else {
			time = null;
		}

		// depth
		if (newJSONObject.containsKey(DEPTH_KEY)) {
			depth = (double) newJSONObject.get(DEPTH_KEY);
		} else {
			depth = null;
		}

		// optional values
		// latitude error
		if (newJSONObject.containsKey(LATITUDE_ERROR_KEY)) {
			latitudeError = (double) newJSONObject.get(LATITUDE_ERROR_KEY);
		} else {
			latitudeError = null;
		}

		// longitude error
		if (newJSONObject.containsKey(LONGITUDE_ERROR_KEY)) {
			longitudeError = (double) newJSONObject.get(LONGITUDE_ERROR_KEY);
		} else {
			longitudeError = null;
		}

		// time error
		if (newJSONObject.containsKey(TIME_ERROR_KEY)) {
			timeError = (double) newJSONObject.get(TIME_ERROR_KEY);
		} else {
			timeError = null;
		}

		// depth error
		if (newJSONObject.containsKey(DEPTH_ERROR_KEY)) {
			depthError = (double) newJSONObject.get(DEPTH_ERROR_KEY);
		} else {
			depthError = null;
		}
	}

	/**
	 * Converts the contents of the class to a json object
	 * 
	 * @return Returns a JSONObject containing the class contents
	 */
	@SuppressWarnings("unchecked")
	public JSONObject toJSON() {

		JSONObject newJSONObject = new JSONObject();

		Double jsonLatitude = getLatitude();
		Double jsonLongitude = getLongitude();
		Date jsonTime = getTime();
		Double jsonDepth = getDepth();
		Double jsonLatitudeError = getLatitudeError();
		Double jsonLongitudeError = getLongitudeError();
		Double jsonTimeError = getTimeError();
		Double jsonDepthError = getDepthError();

		// Required values
		// latitude
		if (jsonLatitude != null) {
			newJSONObject.put(LATITUDE_KEY, jsonLatitude);
		}

		// longitude
		if (jsonLongitude != null) {
			newJSONObject.put(LONGITUDE_KEY, jsonLongitude);
		}

		// depth
		if (jsonDepth != null) {
			newJSONObject.put(DEPTH_KEY, jsonDepth);
		}

		// time
		if (jsonTime != null) {
			newJSONObject.put(TIME_KEY, Utility.formatDate(jsonTime));
		}

		// Optional values
		// latitude error
		if (jsonLatitudeError != null) {
			newJSONObject.put(LATITUDE_ERROR_KEY, jsonLatitudeError);
		}

		// longitude error
		if (jsonLongitudeError != null) {
			newJSONObject.put(LONGITUDE_ERROR_KEY, jsonLongitudeError);
		}

		// depth error
		if (jsonDepthError != null) {
			newJSONObject.put(DEPTH_ERROR_KEY, jsonDepthError);
		}

		// time error
		if (jsonTimeError != null) {
			newJSONObject.put(TIME_ERROR_KEY, jsonTimeError);
		}

		return (newJSONObject);

	}

	/**
	 * Validates the class.
	 * 
	 * @return Returns true if successful
	 */
	public boolean isValid() {
		if (getErrors() == null) {
			return (true);
		} else if (getErrors().size() == 0) {
			return (true);
		} else {
			return (false);
		}
	}

	/**
	 * Gets any validation errors in the class.
	 * 
	 * @return Returns a List&lt;String&gt; of any errors found
	 */
	public ArrayList<String> getErrors() {

		Double jsonLatitude = getLatitude();
		Double jsonLongitude = getLongitude();
		Date jsonTime = getTime();
		Double jsonDepth = getDepth();
		Double jsonLatitudeError = getLatitudeError();
		Double jsonLongitudeError = getLongitudeError();
		Double jsonTimeError = getTimeError();
		Double jsonDepthError = getDepthError();

		ArrayList<String> errorList = new ArrayList<String>();

		// Required Keys
		// latitude
		if (jsonLatitude == null) {
			// latitude not found
			errorList.add("No Latitude in Hypo Class.");
		} else if ((jsonLatitude < -90) || (jsonLatitude > 90)) {
			// invalid latitude
			errorList.add(
					"Latitude in Hypo Class not in the range of -90 to 90.");
		}

		// longitude
		if (jsonLongitude == null) {
			// longitude not found
			errorList.add("No Longitude in Hypo Class.");
		} else if ((jsonLongitude < -180) || (jsonLongitude > 180)) {
			// invalid longitude
			errorList.add(
					"Longitude in Hypo Class not in the range of -180 to 180.");
		}

		// time
		if (jsonTime == null) {
			// time not found
			errorList.add("No Time in Hypo Class.");
		}

		// depth
		if (jsonDepth == null) {
			// depth not found
			errorList.add("No Depth in Hypo Class.");
		} else if ((jsonDepth < -100) || (jsonDepth > 1500)) {
			// invalid depth
			errorList.add(
					"Depth in Hypo Class not in the range of -100 to 1500.");
		}

		// Optional Keys
		// Currently no validation criteria for optional values LatitudeError,
		// LongitudeError, TimeError, and DepthError.

		// success
		return (errorList);
	}

	/**
	 * @return the latitude
	 */
	public Double getLatitude() {
		return latitude;
	}

	/**
	 * @return the longitude
	 */
	public Double getLongitude() {
		return longitude;
	}

	/**
	 * @return the time
	 */
	public Date getTime() {
		return time;
	}

	/**
	 * @return the depth
	 */
	public Double getDepth() {
		return depth;
	}

	/**
	 * @return the latitudeError
	 */
	public Double getLatitudeError() {
		return latitudeError;
	}

	/**
	 * @return the longitudeError
	 */
	public Double getLongitudeError() {
		return longitudeError;
	}

	/**
	 * @return the timeError
	 */
	public Double getTimeError() {
		return timeError;
	}

	/**
	 * @return the depthError
	 */
	public Double getDepthError() {
		return depthError;
	}

}
